packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -0,0 +1,250 @@
+const subhub = require('../subhub/client');
+
+const stripe = require('stripe');
+
+/** @typedef {import('stripe').customers.ICustomer} Customer */
+/** @typedef {import('stripe').products.IProduct} Product */
+/** @typedef {import('stripe').plans.IPlan} Plan */
+/** @typedef {import('stripe').subscriptions.ISubscription} Subscription */
+/** @typedef {import('stripe').customers.ICustomerSubscriptions} Subscriptions */
+
+/**
+ * @typedef AbbrevProduct
+ * @property {string} product_id
+ * @property {string} product_name
+ * @property {Product['metadata']} product_metadata
+ */
+
+/**
+ * @typedef AbbrevPlan
+ * @property {string} plan_id
+ * @property {string} plan_name
+ * @property {Product['metadata']} plan_metadata
+ * @property {string} product_id
+ * @property {string} product_name
+ * @property {Product['metadata']} product_metadata
+ * @property {Plan['interval']} interval
+ * @property {Plan['amount']} amount
+ * @property {Plan['currency']} currency
+ */
+
+/**
+ * Get a cached result at a cache key and regenerated it with `refreshFunction`
+ * if its expired.
+ *
+ * @template T
+ * @param {*} redis
+ * @param {string} cacheKey
+ * @param {number} cacheTtl
+ * @param {() => Promise<T>} refreshFunction
+ * @returns {Promise<T>} possibly cached result
+ */
+async function cachedResult(log, redis, cacheKey, cacheTtl, refreshFunction) {
+  if (cacheTtl) {
+    try {
+      const json = await redis.get(cacheKey);
+      if (json) {
+        return JSON.parse(json);
+      }
+    } catch (err) {
+      log.error(
+        `subhub.cachedResult.${refreshFunction.name}.getCachedResponse.failed`,
+        { err }
+      );
+    }
+  }
+
+  const result = await refreshFunction();
+  if (cacheTtl) {
+    redis
+      .set(cacheKey, JSON.stringify(result), 'EX', cacheTtl)
+      .catch(err =>
+        log.error(
+          `subhub.cachedResult.${refreshFunction.name}.cacheResponse.failed`,
+          { err }
+        )
+      );
+  }
+  return result;
+}
+
+class StripeHelper {
+  /**
+   * Create a Stripe Helper with built-in caching.
+   *
+   * @param {object} log
+   * @param {object} config
+   */
+  constructor(log, config) {
+    this.log = log;
+    this.cacheTtlSeconds = config.subhub.plansCacheTtlSeconds;
+    const redis =
+      this.cacheTtlSeconds &&
+      require('../redis')(
+        {
+          ...config.redis,
+          ...config.redis.subhub,
+        },
+        log
+      );
+    this.plansCacheIsEnabled = this.cacheTtlSeconds && redis;
+
+    this.stripe = new stripe(config.subscriptions.stripeApiKey, {
+      maxNetworkRetries: 3,
+    });
+    this.redis = redis;
+  }
+
+  /**
+   * Fetch all product data and cache it if Redis is enabled.
+   *
+   * Use `allProducts` below to use the cached-enhanced version.
+   *
+   * @returns {Promise<AbbrevProduct[]>} All the products.
+   */
+  async fetchAllProducts() {
+    const products = [];
+    for await (const product of this.stripe.products.list()) {
```

Comment:
TIL [for await of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of).